File: benchmarks/xlang/src/kmeans.lua
local n, k, iters = 200000, 10, 15
local xs, ys = {}, {}
local seed = 42
for i = 1, n do
  seed = seed * 48271 % 2147483647
  xs[i] = 10.0 * seed / 2147483647.0
  seed = seed * 48271 % 2147483647
  ys[i] = 10.0 * seed / 2147483647.0
end
local t0 = os.clock()
local stride = n // k
local cx, cy = {}, {}
for c = 1, k do cx[c] = xs[(c - 1) * stride + 1]; cy[c] = ys[(c - 1) * stride + 1] end
local assign = {}
for it = 1, iters do
  for i = 1, n do
    local best, bd = 1, 1000000.0
    for c = 1, k do
      local dx = xs[i] - cx[c]
      local dy = ys[i] - cy[c]
      local d = dx * dx + dy * dy
      if d < bd then bd = d; best = c end
    end
    assign[i] = best
  end
  local sx, sy, ct = {}, {}, {}
  for c = 1, k do sx[c] = 0.0; sy[c] = 0.0; ct[c] = 0 end
  for i = 1, n do
    local c = assign[i]
    sx[c] = sx[c] + xs[i]; sy[c] = sy[c] + ys[i]; ct[c] = ct[c] + 1
  end
  for c = 1, k do
    if ct[c] > 0 then cx[c] = sx[c] / ct[c]; cy[c] = sy[c] / ct[c] end
  end
end
local sizes = {}
for c = 1, k do sizes[c] = 0 end
for i = 1, n do sizes[assign[i]] = sizes[assign[i]] + 1 end
local largest = 0
for c = 1, k do if sizes[c] > largest then largest = sizes[c] end end
print("CHECK " .. largest)
print("MS " .. math.floor((os.clock() - t0) * 1000))

File: benchmarks/xlang/src/kmeans.js
const n = 200000, k = 10, iters = 15;
const xs = [], ys = [];
let seed = 42;
for (let i = 0; i < n; i++) {
  seed = seed * 48271 % 2147483647;
  xs.push(10 * seed / 2147483647);
  seed = seed * 48271 % 2147483647;
  ys.push(10 * seed / 2147483647);
}
const t0 = performance.now();
const stride = Math.floor(n / k);
const cx = [], cy = [];
for (let c = 0; c < k; c++) { cx.push(xs[c * stride]); cy.push(ys[c * stride]); }
const assign = new Array(n).fill(0);
for (let it = 0; it < iters; it++) {
  for (let i = 0; i < n; i++) {
    let best = 0, bd = 1000000;
    for (let c = 0; c < k; c++) {
      const dx = xs[i] - cx[c], dy = ys[i] - cy[c];
      const d = dx * dx + dy * dy;
      if (d < bd) { bd = d; best = c; }
    }
    assign[i] = best;
  }
  const sx = new Array(k).fill(0), sy = new Array(k).fill(0), ct = new Array(k).fill(0);
  for (let i = 0; i < n; i++) {
    const c = assign[i];
    sx[c] += xs[i]; sy[c] += ys[i]; ct[c]++;
  }
  for (let c = 0; c < k; c++) if (ct[c] > 0) { cx[c] = sx[c] / ct[c]; cy[c] = sy[c] / ct[c]; }
}
const sizes = new Array(k).fill(0);
for (let i = 0; i < n; i++) sizes[assign[i]]++;
console.log(`CHECK ${Math.max(...sizes)}`);
console.log(`MS ${Math.round(performance.now() - t0)}`);
